### ISSUE TITLE
Trailing slashes in `push`-url lead to unauthorized

### ISSUE THREAD

**Issue by mapfel** [fsprojects/Paket#3592]
After several hours of playing around, I give up & hope for assistance here.

Starting point
- Azure DevOps private NuGet Feed  
  https://xxx.pkgs.visualstudio.com/_packaging/SpikeFeed/nuget/v3/index.json
- Created a PAT with permissions "Packaging (Read, write, & manage)" via that site  
  https://xxx.visualstudio.com/_usersSettings/tokens

Now the next steps are unclear for me.

What I've tried:

- `.paket/paket config add-credentials https://xxx.pkgs.visualstudio.com/_packaging/SpikeFeed/nuget/v3/index.json --username PAT --password <token> --verify`
- `.paket/paket.exe config add-token SpikeFeed2 <token> --verify`

both in several variants regarding URL, key, token name:

- `https://xxx.pkgs.visualstudio.com/_packaging/SpikeFeed/nuget/v3/index.json`
- `https://xxx.pkgs.visualstudio.com/_packaging/SpikeFeed/`
- `https://xxx.pkgs.visualstudio.com/_packaging/yyy/nuget/v2/` (from JSON return)
- SpikeFeed vs AzureDevOps vs ...

Later the `paket push` attempts:

- `.paket/paket.exe push --url "https://xxx.pkgs.visualstudio.com/_packaging/SpikeFeed/nuget/v3/index.json" --api-key SpikeFeed2 -v "./temp/Spike.Package.1.0.2.nupkg"`

 end up in different error messages:

- `WebException: The remote server returned an error: (401) Unauthorized.`
- `WebException: The remote server returned an error: (405) Method Not Allowed.`

Hopefully, somebody can give me a hint to get it run.

**Comment by forki** [comment]
Add @isaacabraham. Any recommendations here?

Marko Apfel <notifications@github.com> schrieb am Fr., 7. Juni 2019, 17:38:

**Comment by matthid** [comment]
@mapfel Yes it is a bit tricky to get working but it definitely does. We use it daily on our Azure DevOps 2019 On-Premise instance (and TFS 2017).
Some notes which might or might not help:
- The `--api-key` parameter when pushing is basically ignored. Official Docs use(d) `VSTS`
- Does `--verify` complain or succeed?
- Have you tried using your actual user-name in combination with the token?
- If everything fails you can take a look at the code here: https://github.com/isaacabraham/vsts-fsharp
- I think what might happens is that your current windows-login is used instead of the data you set

Can you provide a way/steps to easily reproduce this?

**Comment by mapfel** [comment]
@matthid Thanks for your assistance. I guess it is a small mistake from me - probably in the understanding how it works and has to be used.

My `paket.config` contains this:

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <credentials>
    <credential source="https://xxx.pkgs.visualstudio.com/_packaging/xxx/nuget/v2" username="marko.apfel@xxx" password="xxx" authType="basic" salt="xxx" />
    <token source="https://xxx.pkgs.visualstudio.com/_packaging/xxx/nuget/v2/" value="xxx" />
  </credentials>
</configuration>
```

btw: currently I try to get it to run on my machine so `paket.config` is in my roaming profile stored. Later `paket push` should be part of the build pipeline. How to handle the `paket.config` then?

Tried in these combinations:
- credential only
- credential and token
- token only

`--verify` succeeded.

Running command

```prompt
.paket/paket.exe push --url https://xxx.pkgs.visualstudio.com --api-key VSTS --endpoint _packaging/xxx/nuget/v2/ temp/Spike.Package.1.0.3.nupkg
```

Are these `--url` and `--endpoint` parameter in the right style?

End up so:
```prompt
Paket version 5.209.1
Pushing package temp/Spike.Package.1.0.3.nupkg to https://xxx.pkgs.visualstudio.com/_packaging/xxx/nuget/v2/ - trial 1
Could not push temp/Spike.Package.1.0.3.nupkg: The remote server returned an error: (401) Unauthorized.
...
Pushing package temp/Spike.Package.1.0.3.nupkg to https://xxx.pkgs.visualstudio.com/_packaging/xxx/nuget/v2/ - trial 5
Performance:
 - Runtime: 2 seconds
Paket failed with
-> WebException: The remote server returned an error: (401) Unauthorized.
```

**Comment by matthid** [comment]
@mapfel I will try around later but ideally it should look like this (all untested):

```
.paket/paket.exe push --source https://pkgs.dev.azure.com/fakebuild/_packaging/ManualTestFeed/nuget/v3/index.json --api-key VSTS temp/Spike.Package.1.0.3.nupkg
```

The sad reality is that we have still https://github.com/fsprojects/Paket/issues/3322 as open issue.
Therefore we need to download the `json` by hand and use the url with the given type:

```
.paket/paket.exe push --source https://pkgs.dev.azure.com/fakebuild/_packaging/04a6f12d-0123-43a5-9d94-52f45c020c1b/nuget/v2/ --api-key VSTS temp/Spike.Package.1.0.3.nupkg
```

Now the credentials might no longer match so we could run

```
.paket/paket config add-credentials https://pkgs.dev.azure.com/fakebuild/_packaging/04a6f12d-0123-43a5-9d94-52f45c020c1b/nuget/v2/ --username PAT --password <token> --verify
```

After this it "might" work ...

**Comment by matthid** [comment]
@mapfel Ok I tested a bit, apparently `add-credentials` removes the trailing slash and `push` is not accounting for that. Therefore

```diff
-.paket/paket.exe push --url https://pkgs.dev.azure.com/fakebuild/_packaging/04a6f12d-0123-43a5-9d94-52f45c020c1b/nuget/v2/ --api-key VSTS temp/Spike.Package.1.0.3.nupkg
+.paket/paket.exe push --url https://pkgs.dev.azure.com/fakebuild/_packaging/04a6f12d-0123-43a5-9d94-52f45c020c1b/nuget/v2 --api-key VSTS temp/Spike.Package.1.0.3.nupkg
```

works. Note the missing trailing slash in the URL... @mapfel Please confirm that this way it works for you as well...

@forki I guess we can consider this a bug report now ;)

**Comment by mapfel** [comment]
@matthid, unfortunately, the same issues.

I did the following:

1. Downloaded the JSON and searched for `...nuget/v2` URL
2. Add credentials via `.paket/paket.exe config add-credentials https://pkgs.dev.azure.com/xxx/_packaging/e0fd9d14-e452-4546-adfa-xxx/nuget/v2/ --username PAT --password <token> --verify  
  <token> from `https://xxx.visualstudio.com/_usersSettings/tokens` (joxqxm3vcp67umw3szlfk24v4...)
3. run `.paket/paket.exe push --url https://pkgs.dev.azure.com/xxx/_packaging/e0fd9d14-e452-4546-adfa-xxx/nuget/v2/ --api-key VSTS temp/Spike.Package.1.0.3.nupkg`

Result as above: `WebException: The remote server returned an error: (401) Unauthorized.`

It would be really great if you can check later this workflow.
Thanks in advance

**Comment by mapfel** [comment]
Yes! @matthid that works!!!

**Comment by matthid** [comment]
> Yes! @matthid that works!!!

I think we could improve the user experience a lot here ;)

By the way (for completeness sake) this is how I created the token:

![image](https://user-images.githubusercontent.com/1866463/59163287-301cd400-8aff-11e9-9146-a2172ecdfbb0.png)

**Comment by mapfel** [comment]
Last question: How I have to proceed for using Azure DevOps build pipelines and handling the `paket.config`??

**Comment by matthid** [comment]
The easiest way is to use https://github.com/isaacabraham/vsts-fsharp and install the corresponding tasks on your DevOps instance.

**Comment by mapfel** [comment]
btw: as promised for the other questions: in my upcoming vacation I will document these topics in details with samples and screenshots and share with the community.

**Comment by mapfel** [comment]
The trailing slash gets lost by saving the URL into `paket.config`. 
Adding the slash manually to the `source` value it will work as expected.
So the algorithm should be tolerated regarding with and without ending backslash.

**Comment by Shmew** [comment]
Not entirely related, but something else I noticed today is that to get azure devops to work properly with paket config you must do add-credentials and the PAT must have full permissions.
